Performance fixes for large repos (responsive grid during loading)

spdr870 opened the pull request:
This is an experimental branch with an attempt to improve performance. It is not stable/finished yet. I mainly want to discuss the changes made so far. The code has not been cleaned, some commented code is still there. The changes where initially done in the 2.51 version, then ported to the 3.00.01 version.

For comparison the Linux repository loads in < 1 second, compared to ~10 seconds (https://github.com/torvalds/linux.git).

Fixes #

Changes proposed in this pull request:
- Do not use topo-order when loading revision log. Its performance is equal to date-order and a lot slower then let git return the log in the order it is in the index.
- Remove locks in the revision graph to improve responsiveness
- Draw revisions as soon as the first commits are loaded, do not wait untill everything is ready
- Stop caching revision graph as soon as visible section is cached to decrease memory usage and not waste cpu cycles. Continue after scrolling down, stop when scrolling up.
 
What has been changed in de UI
- I did remove loading the body text to save some memory, but this can be reverted.
- I removed the loading panel for now. Reason is that I could not find an easy way to remove it on the exact right time.

What needs more work
- (Re)Selecting the revisions. It just doesn't feel right.
- Revision graph column doesn't always resize to the correct width.
- Stability has not been tested enough on version 3.00.01.

What did I do to test the code and ensure quality:
- Several weeks of testing/using the new version.
- Tested with medium sized repositories (500k commits, 5k branches, 5k tags, 2gb index size)

Has been tested on (remove any that don't apply):
- GIT 2.18
- Windows 10

spdr870 commented:
This change is not stable. It needs a lot more testing. It was pretty stable on 2.51, but for 3.00.01 it needs more work.

Personally I need these patches, since version 3.00.00 is too slow. Main reason is that it only renders when the entire revision graph is loaded. For medium sized repositores this is not workable.

RussKie commented:
Yes, I've noticed some performance degradation in v3 on largish repos.

Ok, let's see how fast we can stabilise it.

spdr870 commented:
I will work on it next week, I hope I will get it stable. It would be nice to get some feedback on the approach. Especially because I removed some locks.

RussKie commented:
Hopefully our threading residents @jbialobr, @drewnoakes and @sharwell will find time to review.
It is EOD here, I'll pull the changes tomorrow to test

gerhardol commented:
I got some null exceptions when moving from GE to a submodule
Assuming that this is due to the removed locks

spdr870 commented on GitCommands/RevisionReader.cs:
```diff
@@ -99,8 +99,8 @@ public sealed class RevisionReader : IDisposable
-                /* Commit subject  */ "%s%n%n" +
-                /* Commit body     */ "%b";
+                /* Commit subject  */ "%s%n%n"; // +
+                /////* Commit body     */ "%b";
```
I didn't meassure the performance difference. This can be reverted. If this gained any performance, it is neglectable.

gerhardol replied:
I want to have the Body back, maybe as an option.
Right now the body is missing also in the CommitInfo panel

spdr870 replied:
I reverted the change. 
Is there any way to disable the popup with the commit body? Since there is no timer on the tooltip, it is really annoying when all commits have long commit messages. For example, in the linux repository most commits have over 15 lines causing a lot of popups. A timer would be better in my opinion.  Bit off-topic though....

gerhardol replied:
More OT: The tool tip is great in my opinion, but it must be limited to a few lines of the commit body too in addtion to the timer.

spdr870 commented:
> I got some null exceptions when moving from GE to a submodule
> Assuming that this is due to the removed locks

Probably due to removing the locks indeed. I did test the performance without removing the locks. The results are that the initial loading is just as fast, only when scrolling it gets very sluggish with the locks.

I cherry picked your commit, with some minor tweaks. Movenext should return false if something is failing. The next run will retry. In this case it doesn't really matter, since another thread probably cleared the graph.

Is the performance difference noticable in your environment?

gerhardol commented:
> I cherry picked your commit, with some minor tweaks. 

My commit was just to continue, not reviewed (should be OK, but there may be other similar situations). I got yet another exception due to concurrent access, that should not be merged as is... (I just do not get exception)

> Is the performance difference noticable in your environment?

Big difference just for GE at home. Another more submodule intensive scenario next week, there is a bigger slowdown there.

RussKie commented:
I've run this branch and it definitely feels much snappier.

gerhardol commented:
Index out of bound too (another patch just to continue). Locks may have to be reintroduced.

But the slowdowns in master seen in some mostly submodule heavy repos are almost gone in this PR, most often much faster than 2.x too. I probably want at least parts of this PR in 3.0 release.

spdr870 commented:
I also had to fix an index out of bounds yesterday. Hope the simple bounds check you did is enough, but I am affraid not. Since the grid itself is not threadsafe.

Did you also test with 'Order revision by date' turned off? Thats also a big performance hog.

gerhardol commented:
The patches to index handling etc are very temporary, some threadsafety is needed. But I have enough to test.

In master, some parts are faster and well trimmed. When opening some repos/submodules, the UI hangs though for a number of seconds. This PR mostly fixes that, a few unexpected slowdowns still - that could be the same as in 2.x though. Has deactivated 'Order revision by date' but the repos I work with are not large enough to show a significant difference.

gerhardol commented:
Latest commit creates art:
![image](https://user-images.githubusercontent.com/6248932/46042894-d4e4b600-c116-11e8-8ae3-43c5db93d036.png)

should be
![image](https://user-images.githubusercontent.com/6248932/46042919-e75eef80-c116-11e8-9fa3-d96753f20ea1.png)

spdr870 commented:
@gerhardol Could you verify what commit you where using during the test? And if this 'art' occurred only once, all the time or occasional?

gerhardol commented:
It occurs every time on that repo.
I have rebased, merged remote branches, removed/added remotes several times there, no good explanation.
I saw it first yesterday, so I assumed it was the latest commit as I did not see it before.
(I can investigate in a couple of days or next week, instruction how to recreate or why it occurs.)

spdr870 commented:
@gerhardol  In this commit I did something that might broke it. Espescially the cahnge on line 512-515.
Author:			Henk Westhuis <Henk.Westhuis@ultimo.com>
Date:			14 hours ago (25-9-2018 20:38:28)
Commit hash:	783da68d9732a79ddf98122e2acab034329ed2b5
Child:			d8ceba663e
Parent:			eb3660c53a

Minor threadsafety fixes

Related links: View on GitHub

gerhardol commented:
I have got the following a couple of times, after a merge
(not a clean build of this branch, no known functionality conflicts)

![image](https://user-images.githubusercontent.com/6248932/46263675-60dd5000-c512-11e8-8632-ade28b8fd497.png)

spdr870 commented:
@RussKie I will prepare a better and cleaner pr. Commit eb3660c is part of the pr, asking git to order the revisions (in any way) forces git to walk through the entire history before returning the first result. This is very expensive. In my case, git takes ~5 seconds before when ordering is enabled. In larger repositories this is longer.

spdr870 commented:
> I have got the following a couple of times, after a merge
> (not a clean build of this branch, no known functionality conflicts)
> 
> ![image](https://user-images.githubusercontent.com/6248932/46263675-60dd5000-c512-11e8-8632-ade28b8fd497.png)
> 
> Scrolling to the top:
> 
> ![image](https://user-images.githubusercontent.com/6248932/46263692-9d10b080-c512-11e8-9d1b-5d4d8bd2e57b.png)
> 
> (refresh button resolves)

I sometimes have the same, but if I scroll down/up it draws correcly. Can you confirm if this is also the case with your issue? If this is the case, its only a redraw issue. If not, it might be that the graph failed to build.

gerhardol commented:
> if I scroll down/up it draws correcly. Can you confirm if this is also the case with your issue? If this is the case, its only a redraw issue. If not, it might be that the graph failed to build.

scrolling do not fix the graph

vbjay commented:
My thoughts are that it is because of the lock removal.

On Mon, Oct 1, 2018, 4:36 PM Gerhard Olsson <notifications@github.com>
wrote:

> if I scroll down/up it draws correcly. Can you confirm if this is also the
> case with your issue? If this is the case, its only a redraw issue. If not,
> it might be that the graph failed to build.
>
> scrolling do not fix the graph
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/5473#issuecomment-426053538>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADdhsfHY6694ysk1lPKkFOwlS3IBQEEIks5ugny9gaJpZM4W1r2c>
> .
>

spdr870 commented:
@gerhardol @vbjay I reenabled the locks on branch feature/graphperformance_3.00.00_locks. It is pushed. Not sure if this fixes your problems, but its worth a try. Putting the locks back doesn't affect the initial loading performance. There are other fixes for that. It mainly affect scrolling down when the graph is not completely build yet.

gerhardol commented:
I still see the "art".
The locks or a rewrite is very likely needed and some way to avoid the "art" too. 
I have not investigated why, I just report my findings when trying out the branch.
The improvements in it are much appreciated. I still see occasional freezes, but much fewer and shorter than with master.

![image](https://user-images.githubusercontent.com/6248932/46381896-987e0080-c6a8-11e8-9fdc-b9e68af37885.png)

spdr870 commented:
I do not believe the 'art' is caused by locks. The graphmodel expects the nodes to be processed in topo order. I disabled topo-order. When you select date-order, it also forces it to be in topo-order. There is even a guard for this (GraphModel.cs, line 720). I expect this code to be hit in your case, only, but it should be able to repair itself.

Problem with topo-order is that it is just too slow.

spdr870 commented:
@gerhardol I was finally able to simulate the 'art'. I do not get it every time, but I was able to debug it. It looks like an issue that occurred because the collections used are not threadsafe. I replaced the used lists with threadsafe variants. After this, I could not simulate the issue again. Hope it helps! 

I used SynchronizedCollection and ConcurrentDictionary. The SynchronizedCollection is no longer recommended to be used, because it uses locks instead of SpintWiat. But I choose this so I could easily replace the existing lists.

RussKie commented on GitUI/UserControls/RevisionGrid/Graph/GraphModel.cs:
```diff
@@ -746,7 +757,15 @@ public void Prune()
+                    }
+
+                    _nodeByObjectId.TryRemove(node.ObjectId, out removedValue);
+                }
```
I'm having troubles understanding these `if`s.
Could you please clarify why `TryAdd` and `TryRemove` are duplicated?

spdr870 replied:
Removed

spdr870 commented on GitUI/UserControls/RevisionGrid/Graph/GraphModel.cs:
```diff
@@ -288,7 +289,17 @@ bool GetOrCreateNode(ObjectId objectId, out Node n)
+                        if (Debugger.IsAttached)
+                        {
+                            Debugger.Break();
+                        }
```
Just read the documentation, you are right. It can only fail if:
- and argument exception occurs
- overflow exception occurs
- when the key already exists

I removed it.

spdr870 commented:
Is it possible to archive the entire repo (just the .git folder) and send it to me? Do not garbage collect, the fragmentation in the repository could be the cause.

gerhardol commented:
> Is it possible to archive the entire repo (just the .git folder) and send it to me?

I will, as my (brief) debugging has not led anywhere.

spdr870 commented:
@gerhardol I recieved and tested the repository. I cannot reproduce the 'art' on this repository. I tried git versions 2.18 and 2.19.
Could you maybe run this command on the repo and send me the output? I wonder what the output is, and if it is different than the output I am getting. Also, I could test with your output.
`git -c log.showSignature=false log -z --pretty=format:"%H%T%P%n%at%n%ct%n%e%n%aN%n%aE%n%cN%n%cE%n%s%n%n" --all --boundary --not --glob=notes --not --max-count=200000 -- > test.txt`

gerhardol commented:
I can reproduce with the repo on another PC (not a clean build of your changes, some other changes too that should be in separate files).
Sent the output

gerhardol commented:
Further observation: I only see the 'art' when displaying reflog in the grid

spdr870 commented:
Ok, the information about reflog makes sense! Now I understand. I can also simulate it now. Thanks for the testing! According to the documentation:
- git log returns the HEAD revision and traverses to parents
- git reflog returns the info from the reflog. Not in order. 

So if reflog is used, we need to ask git to return the revisions in topo order. If reflog is not used, this is not needed. Ok, this is easy to fix! I will make a patch that if reflog is enabled, topo order is forced.

But, this means that:
- If reflog is used, the loading is slower
- If date-order is used, the loading is slower
How much slower depends on the repo size. In my case, you really want the fastest setup as possible. I think the default settings should perform as fast as possible.

spdr870 commented:
I added a commit that re-enables a loading indicator. It is more subtle, since the revisions are shown before loading is completed. This allows to already start working before loading is complete. When the 'slow' options are used (date-order or reflog) the loading indicator is needed.

gerhardol commented on GitCommands/RevisionReader.cs:
```diff
@@ -162,8 +162,8 @@ ArgumentBuilder BuildArguments()
-                    { AppSettings.OrderRevisionByDate, "--date-order", "--topo-order" },
-                    { AppSettings.ShowReflogReferences, "--reflog" },
+                    { AppSettings.OrderRevisionByDate, "--date-order" },
+                    { AppSettings.ShowReflogReferences, "--reflog --topo-order" }, // if reflog is used, the revisions are not returned in topo-order. Force topo order, since we require topo order for the revision graph.
```
Should this be visible in settings, so date-order is set when reflog is set?

spdr870 replied:
Its topo order. The topo-order means that git.exe will rearrange the order of the commits in how they are returned from the git log command. Topo order forces a specific order in parent traversal. Not sure how this exaclty works. Normally git log will return all commits in child->parent order. Adding topo-order also orderes between branches. This can be usefull.

In case of --reflog, the log is used to determine the order of the commits to be returned. When you rebase often, the log is not always in the order of child->parent. This is a problem because our graph algorithm requres a child->parent order.

So to be short; not sure what we should make visible in the settings. I do agree that for advanced users it is helpfull to know what a setting does exactly. I think the same applies about settings that slow down the applications. For example, reflog and date-order slow down the application in some cases. Shouldn't we warn the user is some way?

gerhardol replied:
We should have a wikipage with performance suggestions and have reasonable defaults
(maybe a link from settings)

A gold plated solution could be a popup if an operation takes too long time, I vote no

gerhardol commented on GitUI/UserControls/RevisionGrid/RevisionDataGridView.cs:
```diff
@@ -218,7 +218,6 @@ public IReadOnlyList<ObjectId> SelectedObjectIds
                 }
 
                 lock (_backgroundEvent)
-                lock (_graphModel)
```
Is removing the locks the way to go?
Anything else to consider related to this?

spdr870 replied:
It should be safe:
- One thread is adding items.
- No thread is removing items.
- Multiple threads are reading.
- The lists/collections are now all of threadsafe types (sycnhronizedcollection & concurrentdictionary). This should make sure the count is reliable.
- Before the model is cleared, the other threads are killed.

gerhardol commented on GitUI/UserControls/RevisionGrid/RevisionGridControl.cs:
```diff
@@ -1069,7 +1090,8 @@ int SearchRevision(ObjectId objectId)
                 }
 
-                // Not found, so search for its parents
+                /* The code below is extremely slow, TryGetParents will return a huge list of parents
```
What is the consequence of this?
Any risk to miss parents now?

spdr870 replied:
No/yes -> This causes the related commits in the file history not to be drawn in a color. All lines in the graph are grey if the top commit is not the current commit. This is indeed regression; but the loading of the history, in my case, changes from 30sec to 1sec. Not sure what to do with this.

gerhardol replied:
This could be an option (even if there are too many options already). 
I suggest this option is enabled by default though, listed as a tweak with certain effects on a wiki page or so.

spdr870 commented:
When the branch is stable, I will create a proper pull request. There are some changes not required and some code needs to be cleaned up. For example the body is now not loaded, but this should be reenabled. I did some more performance fixes in my fork, but more separation is possible. Lets do some testing and then decide if it is stable. My collegeas and I are testing for several weeks and most problems seem to be ironed out. No recent crashes.

I think we need to make a decision as collaborators/maintainers of this project. What is more important, enable all features or ensure good performance out of the box. Personally I think that users will stop using the program when the performance is bad. I tried several user interfaces for git. Most perform very badly. A few (like sublime merge or gitk) are extremely fast. When the performance is bad, I do not even care looking at the features. When a program is not snappy, it becomes frustrating very soon. This is just my opinion though.

gerhardol commented on GitUI/UserControls/RevisionGrid/RevisionGridControl.cs:
```diff
@@ -1234,7 +1257,7 @@ private void OnGridViewMouseDown(object sender, MouseEventArgs e)
                 case MouseButtons.XButton1: NavigateBackward(); break;
                 case MouseButtons.XButton2: NavigateForward(); break;
-                case MouseButtons.Left when _maximizedColumn != null && _lastVisibleResizableColumn != null:
+                /*case MouseButtons.Left when _maximizedColumn != null && _lastVisibleResizableColumn != null:
```
Why remove this?

gerhardol commented on Plugins/DeleteUnusedBranches/DeleteUnusedBranchesForm.cs:
```diff
@@ -163,18 +163,17 @@ private void Delete_Click(object sender, EventArgs e)
                 await TaskScheduler.Default.SwitchTo(alwaysYield: true);
 
-                if (remoteBranches.Count > 0)
+                foreach (var remoteBranch in remoteBranches)
```
This is probably a separate PR. Any issue?

spdr870 replied:
Agree. Needed this for testing. Issue was->1 branch could not be deleted for some reason. This resulted in none of the 1500 branches I wanted to delete was deleted. Didn't recieve an error, didn't know why it broke. This has nothing to do with this pull request though..:)

RussKie commented:
I think for the most part right now the app lacks in the UI department more than anything. 
I also think we're the only one building a true FOSS git UI client without any financial backing. Our competitors are either commercial or backed by serious finance. And yet a lot of them suck...

Back to the point, we should not compromise on performance, nor we should make the UX more complex than necessary. We need to simplify configurations and make things easier to discover as much as we can. That's probably the main feedback I received from engineers I work with.

spdr870 commented:
I think that I can make this asynchronous.  To be honest,  I miss this
feature... making it asynchronous makes sure the file history is loaded
fast, and the branch highlights are there as soon as this is ready.  I will
have some time this weekend. I will prepare a merge request first for the
main performance improvements, without this. Then submit 3 separate pr's
for more improvements.
1. The fix for obsolete branch plugin
2. Performance fix for selecting g initial commit (currently on my fork)
3. This feature asynchronous

Op za 6 okt. 2018 15:01 schreef Gerhard Olsson <notifications@github.com>:

> *@gerhardol* commented on this pull request.
> ------------------------------
>
> In GitUI/UserControls/RevisionGrid/RevisionGridControl.cs
> <https://github.com/gitextensions/gitextensions/pull/5473#discussion_r223181853>
> :
>
> > @@ -1069,7 +1090,8 @@ int SearchRevision(ObjectId objectId)
>                      return exactIndex;
>                  }
>
> -                // Not found, so search for its parents
> +                /* The code below is extremely slow, TryGetParents will return a huge list of parents
>
> This could be an option (even if there are too many options already).
> I suggest this option is enabled by default though, listed as a tweak with
> certain effects on a wiki page or so.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/5473#discussion_r223181853>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AACXE_2kdwWR4xxZxJf0kn-nuQUgZs5rks5uiKm8gaJpZM4W1r2c>
> .
>

spdr870 commented:
I rebased and cleaned up this pull request. It only contains the necessary fixes to make the grid responsive. I removed some guards that I and @gerhardol added, since they should not be needed anymore because we use threadsafe collections now.

This PR is ready.

gerhardol commented:
Great. I want to try this branch in active use for a couple of days before approving myself.
I encourage all to test, especially @gitextensions/git-extensions-source to try this too, I suggest this is merged before next public build.
No real comments on the code but I have not tried to dissect the grid code changes.

In addition to my normal repos, I tried the Linux kernel to get a big repo. (that itself has problems, see bottom.)

The Loading... message appear in the upper right corner and is maybe a little to subtle. But the indicator is hardly seen.
![image](https://user-images.githubusercontent.com/6248932/46580674-9c53af00-ca29-11e8-8720-827e4e2c2e12.png)

A few times (once with the latest build too) I have seen longer startup times like a minute without response with the Linux repo. But such lockups are much fewer than in master and probably similar to 2.x.

In 2.x the display could be distorted in many ways in a large repo. In this branch I have primariily seen the following. These are not blocking merge (and they are likely not regressions) but I mention them for other testers.

Scrolling is much better, but not perfect in a mayor repo.

Many branches are handled well. In this case the lanes could expand some more to see all branches.

![image](https://user-images.githubusercontent.com/6248932/46580579-bee4c880-ca27-11e8-9c44-51d06a62efc9.png)

If you scroll with the scrollbar some down (i.e. not scrolling continuously), the lanes disappear

![image](https://user-images.githubusercontent.com/6248932/46580763-a70f4380-ca2b-11e8-9951-fc1d6a6b18a6.png)

When scrolling or selecting a commit without lanes, they are normally drawn. Occasionally, I see the following though. It is probably auto-corrected after some minutes

![image](https://user-images.githubusercontent.com/6248932/46580813-3caad300-ca2c-11e8-9003-d320a85de529.png)

--
Linux kernel: cannot be checked out in Windows, as the basename "aux" is reserved in Windows and aux.c for instance will fail a checkout. But you can still use the repo to test grid performance. Also note that FullPathResolver.cs Path.Combine() enforces the "device file name" so Uri()  fails with an exception in the diff viewer. A try-catch is probably not worth it for a repo that will not work in Windows anyway.

spdr870 commented:
@gerhardol Good that you tested with the linux kernel repository! For best comparison, test with date-order and reflog turned off. Then you can see the real performance gain. The 3.0 is much faster then.

gerhardol commented:
> @gerhardol Good that you tested with the linux kernel repository! For best comparison, test with date-order and reflog turned off. Then you can see the real performance gain. The 3.0 is much faster then.

Yes, I tested with reflogs off.
If none has a better idea, I will add a wiki page with this information, as well as commit-count limitations.

Measuring UI performance is hard to test and differences between GE versions can be very subjective. This for sure improves the experience.

spdr870 commented:
Fixed the issue where the graphcolumn width is not updated correctly. When scrolling down and the graph is not ready yet, it will be small. It will update when the graph is shown.

spdr870 commented:
Just fixed an issue that the revision grid didn't show any commit when graph is disabled in settings. This isn't regression from this PR though, doesn't work in master either.

spdr870 commented:
Rebased on master

gerhardol commented on GitUI/UserControls/RevisionGrid/Columns/GraphColumnProvider.cs:
```diff
@@ -424,7 +425,6 @@ public override void Clear()
             lock (_graphModel)
             {
                 _graphModel.Clear();
-                ////_graphDataCount = 0;
```
Why keep this comment?

spdr870 replied:
I didn't keep it, I removed id.

jbialobr commented on GitUI/UserControls/RevisionGrid/RevisionDataGridView.cs:
```diff
@@ -497,41 +494,30 @@ void UpdateGraph(int fromIndex, in int toIndex)
-                        // Cache the next item
-                        if (!_graphModel.CacheTo(rowIndex))
+                        Debug.WriteLine("Cached item FAILED {0}", rowIndex);
+                        lock (_backgroundThread)
```
What is the purpose of locking on the `_backgroundThread`? It seems to be suspicious to lock on a reference that may be modified.

spdr870 replied:
These locks can be removed. Done.

spdr870 commented:
> Minor: The Loading indicator could be clearer.

I agree. Graphic design is not my strongest. Also tried the WaitPointer mouse cursor. Hated it.

jbialobr commented on GitUI/UserControls/RevisionGrid/VisibleRowRange.cs:
```diff
@@ -21,7 +21,7 @@ public VisibleRowRange(int fromIndex, int toIndex)
             ToIndex = toIndex;
         }
 
-        public int Count => ToIndex - FromIndex + 1;
```
It should be corrected at the place where VisibleRowRange is created (line 534).

```csharp
            if (toIndex >= _graphModel.Count)
            {
                toIndex = _graphModel.Count - 1; <-- here
            }

            if (toIndex >= 0)
            {
                _visibleRowRange = new VisibleRowRange(fromIndex, toIndex);

```

spdr870 replied:
I'm not sure if I agree.
Example:
FromIndex = 0
ToIndex = 1

Then the count should be 2 (row 0 and 1). From/to are effectually from/to and including. Or do I miss something?

jbialobr replied:
I don't understand what you mean. The VisibleRowRange is meant to include its border values and as you wrote for the range <0,1> the count has to be 2. While after your change the count is 1 (1-0).
The problem is in the place when the toIndex is computed as is assigned with _graphModel.Count.
_graphModel count can not be assigned to toIndex, it has to be decreased by one in order be in the range of the 0-based collection.

spdr870 replied:
Your right. I Misread the diff. I removed the '+1' and not added it. My mistake. I will put bach the +1 in VisibleRowRange. Then it is 1(to)+0(from)+1 == 2 again. Thanks!

spdr870 commented on GitUI/UserControls/RevisionGrid/RevisionDataGridView.cs:
```diff
@@ -423,70 +410,64 @@ private void SetRowCount(int count)
-                            }
-
-                            UpdateGraph(curCount, scrollTo);
+                            this.InvokeAsync(NotifyProvidersVisibleRowRangeChanged).FileAndForget();
```
Agree, done

jbialobr commented on GitUI/UserControls/RevisionGrid/RevisionDataGridView.cs:
```diff
@@ -554,41 +521,44 @@ void UpdateRow(int row)
+
+            if (fromIndex >= _graphModel.Count)
+            {
+                fromIndex = _graphModel.Count;
```
After reverting VisibleRowRange.Count you have to correct this to be 
```csharp
fromIndex = _graphModel.Count - 1;
```
and in the 534 line as well.

spdr870 replied:
Done.

RussKie commented on GitUI/UserControls/RevisionGrid/RevisionGridControl.cs:
```diff
@@ -131,7 +131,20 @@ public RevisionGridControl()
+            {
+                Location = new Point(0, 0),
+                Dock = DockStyle.Right,
+                Padding = new Padding(7, 5, 5, 5),
```
I suspect this needs to be scaled with `DipUtil`.

spdr870 replied:
I copied this from QuichSearchProvider.cs (line 30-40). I attempted replace this with a transparent spinner in the top-right corner, but failed to get it transparent. I do plan to make the loading indicator better. Will do it in a seperate pull request if that's ok with you. I also would like a more present loading indicator for the period that no row is returned from git.exe yet.

RussKie commented:
No worries, I'm good with that.

On Thu, 11 Oct 2018 at 20:55, Henk Westhuis <notifications@github.com>
wrote:

> *@spdr870* commented on this pull request.
> ------------------------------
>
> In GitUI/UserControls/RevisionGrid/RevisionGridControl.cs
> <https://github.com/gitextensions/gitextensions/pull/5473#discussion_r224386807>
> :
>
> > @@ -131,7 +131,20 @@ public RevisionGridControl()
>              InitializeComponent();
>              InitializeComplete();
>
> -            _loadingImage = new LoadingControl();
> +            _loadingControl = new Label
> +            {
> +                Location = new Point(0, 0),
> +                Dock = DockStyle.Right,
> +                Padding = new Padding(7, 5, 5, 5),
>
> I copied this from QuichSearchProvider.cs (line 30-40). I attempted
> replace this with a transparent spinner in the top-right corner, but failed
> to get it transparent. I do plan to make the loading indicator better. Will
> do it in a seperate pull request if that's ok with you. I also would like a
> more present loading indicator for the period that no row is returned from
> git.exe yet.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/5473#discussion_r224386807>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEMyXpJdXopRyOS298dm8MDPALz_-aDCks5ujxWEgaJpZM4W1r2c>
> .
>

gerhardol commented:
great!
